pkg/expansion/system.go

```diff
@@ -40,12 +51,22 @@ func keyForTemplate(template *expansionunversioned.ExpansionTemplate) string {
 func (s *System) UpsertTemplate(template *expansionunversioned.ExpansionTemplate) error {
 	s.lock.Lock()
 	defer s.lock.Unlock()
+	log.V(1).Info("upserting ExpansionTemplate", "template name", template.GetName())
 
 	if err := ValidateTemplate(template); err != nil {
 		return err
 	}
 
-	s.templates[keyForTemplate(template)] = template.DeepCopy()
+	k := keyForTemplate(template)
+	if oldTemp, exists := s.templates[k]; exists {
+		s.graph.removeTemplate(oldTemp)
+	}
+
+	if err := s.graph.addTemplate(template); err != nil {
```

Comment:
If we have an error adding a template due to it creating a cycle, we probably want to also disable the conflicting template(s), to avoid a "first-past-the-post" race condition, where different pods have different opinions as to which expansions are active for a given config.

This begs the question of if it's possible to find a conflicting template? Or at least come up with a definition of "conflicting template" that yields a stable solution after all templates are reconciled. I'm not 100% sure that's a solvable problem.

In which case, how do we reach/define a convergent state for the system of pods as a whole? Disable expansion entirely?

Reply:
Interesting scenario. 

My thinking is that _ideally_ we don't have to disable the entire expansion system --just the conflicting templates (i.e. templates that make up that cycle). In the steady state, once each Pod has reconciled all templates, the systems should be consistent. 

However, enforcing this behavior is a little tricky with how the code is currently set up. When we try to upsert a cyclic template, we basically throw it away (from the expansion system's point of view). I believe mutation is different because it keeps the conflicting mutator in the system, but marks it as conflicting. 

It should be do-able change the expansion system to work more like mutator's `schemaDB`, but is it worth it? I think the value add is small compared to just disabling the entire system whenever we detect a conflict. 

The only value I see partial disablement bringing is in a scenario where someone adds a new `ExpansionTemplate` to their cluster and walks away. In this case, only "part" of their expansion system would be broken. But realistically, I think most people would inspect the resource afterwords to make sure it was created properly.

Reply:
Afterthought: does mutation break down under the same race condition? In other words, if mutators `A` and `B` conflict, and `A` was added first, do we continue to enforce `A` after we add `B`? Reading the code, I'm not sure if schema conflicts are recorded for only the new conflicting mutator, or for the existing mutator(s) that it conflicts with as well ([does `s.Add` add conflicts for the existing mutator too?)](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/mutation/schema/schema.go#L109)).